File: backend/app/routes/task.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.main import get_session
from app.db.models.enums import ProjectPriority, ProjectStatus
from app.db.models.user import User
from app.routes.auth import get_current_user
from app.schemas.task import TaskCreate, TaskResponse, TaskUpdate
from app.services.task_service import TaskServiceDep

router = APIRouter(tags=["tasks"])


@router.post(
    "/projects/{project_id}/tasks",
    response_model=TaskResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_task(
    project_id: UUID,
    task_create: TaskCreate,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_session)],
    task_service: TaskServiceDep,
):
    """
    Create a new task in the project
    """
    return await task_service.create_task(project_id, task_create, current_user)


@router.get("/projects/{project_id}/tasks", response_model=list[TaskResponse])
async def list_tasks(
    project_id: UUID,
    task_service: TaskServiceDep,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_session)],
    status: ProjectStatus | None = None,
    priority: ProjectPriority | None = None,
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=100),
):
    return await task_service.get_tasks_by_project(
        project_id, current_user, status, priority, page, limit
    )


@router.get("/tasks/{task_id}", response_model=TaskResponse)
async def read_task(
    task_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_session)],
    task_service: TaskServiceDep,
):
    task = await task_service.get_task_by_id(task_id, current_user)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.patch("/tasks/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: UUID,
    task_update: TaskUpdate,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_session)],
    task_service: TaskServiceDep,
):
    task = await task_service.update_task(task_id, task_update, current_user)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.delete("/tasks/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    task_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_session)],
    task_service: TaskServiceDep,
):
    success = await task_service.delete_task(task_id, current_user)
    if not success:
        raise HTTPException(status_code=404, detail="Task not found")

File: frontend/src/services/task.ts
import { client } from "@/api/client";
import type { TaskCreateValues, TaskUpdateValues } from "@/schemas/task";
import type { TaskResponse } from "@/types/task";

export const taskService = {
  /**
   * Create a new task in a project
   */
  create: async (
    projectId: string,
    data: TaskCreateValues,
  ): Promise<TaskResponse> => {
    const res = await client.post<TaskResponse>(
      `/projects/${projectId}/tasks`,
      data,
      {
        withCredentials: true,
      },
    );
    return res.data;
  },

  /**
   * Get all tasks for a project
   */
  getByProject: async (
    projectId: string,
    params?: {
      status?: string;
      priority?: string;
      page?: number;
      limit?: number;
    },
  ): Promise<TaskResponse[]> => {
    const res = await client.get<TaskResponse[]>(
      `/projects/${projectId}/tasks`,
      {
        params,
        withCredentials: true,
      },
    );
    return res.data;
  },

  /**
   * Get task by ID
   */
  getById: async (taskId: string): Promise<TaskResponse> => {
    const res = await client.get<TaskResponse>(`/tasks/${taskId}`, {
      withCredentials: true,
    });
    return res.data;
  },

  /**
   * Update task
   */
  update: async (
    taskId: string,
    data: TaskUpdateValues,
  ): Promise<TaskResponse> => {
    const res = await client.patch<TaskResponse>(`/tasks/${taskId}`, data, {
      withCredentials: true,
    });
    return res.data;
  },

  /**
   * Delete task
   */
  delete: async (taskId: string): Promise<void> => {
    await client.delete(`/tasks/${taskId}`, {
      withCredentials: true,
    });
  },
};
